cmd/entrypoint/subcommands/subcommands.go

```diff
@@ -0,0 +1,60 @@
+package subcommands
+
+import (
+	"fmt"
+)
+
+type SubcommandSuccessful struct {
+	message string
+}
+
+func (err SubcommandSuccessful) Error() string {
+	return err.message
+}
+
+type SubcommandError struct {
+	subcommand string
+	message    string
+}
+
+func (err SubcommandError) Error() string {
+	return fmt.Sprintf("%s error: %s", err.subcommand, err.message)
+}
+
+// Process takes the set of arguments passed to entrypoint and executes any
+// subcommand that the args call for. An error is returned to the caller to
+// indicate that a subcommand was matched and to pass back its success/fail
+// state. The returned error will be nil if no subcommand was matched to the
+// passed args, SubcommandSuccessful if args matched and the subcommand
+// succeeded, or any other error if the args matched but the subcommand failed.
+func Process(args []string) error {
+	if len(args) == 0 {
+		return nil
+	}
+	switch args[0] {
+	case CopyCommand:
+		// If invoked in "cp mode" (`entrypoint cp <src> <dst>`), simply copy
+		// the src path to the dst path. This is used to place the entrypoint
+		// binary in the tools directory, without requiring the cp command to
+		// exist in the base image.
+		if len(args) == 3 {
+			src, dst := args[1], args[2]
+			if err := cp(src, dst); err != nil {
+				return SubcommandError{subcommand: CopyCommand, message: err.Error()}
+			}
+			return SubcommandSuccessful{message: fmt.Sprintf("Copied %s to %s", src, dst)}
+		}
+	case DecodeScriptCommand:
+		// If invoked in "decode-script" mode (`entrypoint decode-script <src>`),
+		// read the script at <src> and overwrite it with its decoded content.
+		if len(args) == 2 {
+			src := args[1]
+			if err := decodeScript(src); err != nil {
+				return SubcommandError{subcommand: DecodeScriptCommand, message: err.Error()}
+			}
+			return SubcommandSuccessful{message: fmt.Sprintf("Decoded script %s", src)}
+		}
+	default:
```

Comment:
does this mean invalid args are ignored? (or does this just stop it from processing something that doesnt look like a subcommand?)

Reply:
Yeah, just trying to capture the fact that the default operation is to "do nothing, return nil".